### PR TITLE
Axis-weighted Hausdorff distance

### PR DESCRIPTION
**Pull request by mateuszbaran**
I've added a possibility to make axis-weighted Hausdorff distance. It required very little tweaking :slightly_smiling_face:

### CODE REVIEW

**Review by johnnychen94 — APPROVED** [review]
LGTM but I recommend having @juliohm check the background since he's the expert here.

**Comment by juliohm** [comment]
@mateuszbaran thanks for the contribution. Can you give some reference or share some examples of what the weighted version does? I've never heard of it in the literature.

**Comment by mateuszbaran** [comment]
I have 3D binary images such that one of the axes is sampled at a different interval than other. I think I can just use weighting instead of resampling the image to make all axes sampled at the same interval. I don't have any references for that though.

**Comment by juliohm** [comment]
Got it. I think we need to update the docstrings to reflect these options.
Also the current docstrings are full of latex code and I think this could
be erased in favour of a more succinct docstrings. These equations could go
somewhere else in the docs since they don't render correctly in the
terminal.

**Comment by mateuszbaran** [comment]
I've updated the docstring for `GenericHausdorff` (that's the only thing touched by this PR). For the latex code, my general preference is to use Unicode symbols where possible (for example  ℳ instead of `\mathcal{M}` ) but that's a separate issue. This is more readable and usually still renders correctly in Documenter.jl.

**Comment by juliohm** [comment]
Yes that would be awesome. Maybe we could address this docstrings issue in
a separate PR. I'll review the code carefully again soon.

File: src/hausdorff.jl [hunk 1]
```diff
@@ -79,14 +81,16 @@ Dubuisson, M-P; Jain, A. K., 1994. *A Modified Hausdorff Distance for Object-Mat
...
-const ModifiedHausdorff = GenericHausdorff{MeanReduction, MaxReduction}
-ModifiedHausdorff() = ModifiedHausdorff(MeanReduction(), MaxReduction())
+const ModifiedHausdorff = GenericHausdorff{MeanReduction, MaxReduction, Nothing}
+ModifiedHausdorff() = ModifiedHausdorff(MeanReduction(), MaxReduction(), nothing)
 
 # convert binary image to its distance transform
-hausdorff_transform(img::AbstractArray{Bool}) = distance_transform(feature_transform(img))
-function hausdorff_transform(img::GenericGrayImage)
+function hausdorff_transform(d::GenericHausdorff, img::AbstractArray{Bool})
+    return distance_transform(feature_transform(img, d.transform_weights), d.transform_weights)
```
Comment (juliohm) [inline src/hausdorff.jl:89]:
```suggestion
    return distance_transform(feature_transform(img, d.weights), d.weights)
```

File: test/hausdorff.jl [hunk 2]
```diff
@@ -49,4 +49,30 @@ type_list = generate_test_types([Bool], [Gray])
         end
     end
     test_cross_type(dist, A, B, type_list)
+
+    dist_2 = GenericHausdorff(ImageDistances.MaxReduction(), ImageDistances.MaxReduction(), (1.0, 1.5))
+    dist_3 = GenericHausdorff(ImageDistances.MaxReduction(), ImageDistances.MaxReduction(), (1.0, 1.5, 2.0))
```
Comment (juliohm) [inline test/hausdorff.jl:54]: What about exporting a new constructor:

```julia
AxisWeightedHausdorff(weights) = GenericHausdorff(MaxReduction(), MaxReduction(), weghts)
```

**Review by juliohm — CHANGES_REQUESTED** [review]
Sorry for the delay reviewing this. I completely forgot about it.

I've shortened the long variable names and suggested to maybe export a new variation for convenience. Please let me know what you think. I think we don't want users dealing with
internal names such as GenericHausdorff. Instead they should be using `Hausdorff`, `ModifiedHausdorff` and the possibly new `AxisWeightedHausdorff`.

File: test/hausdorff.jl (hunk 2 above)
Comment (mateuszbaran) [reply test/hausdorff.jl:54]: This is a good idea, I'll add this constructor.

**Review by juliohm — APPROVED** [review]

**Comment by juliohm** [comment]
Thanks @mateuszbaran for these great improvements.

**Pull request merged**